xarray/backends/netCDF4_.py

```diff
@@ -27,21 +25,29 @@
                   '|': 'native'}
 
 
-class BaseNetCDF4Array(NdimSizeLenMixin, DunderArrayMixin):
+class BaseNetCDF4Array(indexing.NDArrayIndexable):
     def __init__(self, variable_name, datastore):
         self.datastore = datastore
         self.variable_name = variable_name
 
         array = self.get_array()
-        self.shape = array.shape
+        self._shape = array.shape
 
         dtype = array.dtype
         if dtype is str:
             # use object dtype because that's the only way in numpy to
             # represent variable length strings; it also prevents automatic
             # string concatenation via conventions.decode_cf_variable
             dtype = np.dtype('O')
-        self.dtype = dtype
+        self._dtype = dtype
+
+    @property
+    def shape(self):
+        return self._shape
```

Comment:
Why are all the private methods / public properties needed?

Reply:
This is because `NDArrayMixin` (the parent of `NDArrayIndexable`) requires this property.
We can avoid this by using the double inheritance... but I think the above is simpler.

Reply:
Ah yes, I've had this problem in python before. Agree this is a better compromise than double inheritance